main/modes/mode_manual.c

```diff
@@ -0,0 +1,400 @@
+{
+    node_t* firstPage = NULL;
+    page_t* newPage = NULL;
+    markdownContinue_t* nextPos = NULL;
+
+    do
+    {
+        newPage = malloc(sizeof(page_t));
+        newPage->chapter = manual->chapter;
+
+        // copyContinue will return a NULL if passed a NULL
+        newPage->mdPos = copyContinue(nextPos);
+        push(&manual->pages, newPage);
+
+        if (firstPage == NULL)
+        {
+            firstPage = manual->pages.last;
+        }
+    }
+    while (drawMarkdown(NULL, markdown, &manual->mdParams, &nextPos, true));
+
+    if (nextPos != NULL)
+    {
+        free(nextPos);
+    }
+
+    return firstPage;
+}
+
+void manualLoadText(bool reverse)
+{
+    ESP_LOGD("Manual", "Loading text");
+    if (manual->markdown != NULL)
+    {
+        freeMarkdown(manual->markdown);
+        manual->markdown = NULL;
+    }
+
+    if (manual->text != NULL)
+    {
+        free(manual->text);
+        manual->text = NULL;
+    }
+
+    manual->text = loadTxt(manual->chapter->file);
+    if (manual->text == NULL)
+    {
+        manual->text = "# Could not load text :(";
+    }
+
+    manual->markdown = parseMarkdown(manual->text);
+
+    // if the page we're trying to use hasn't been created yet, we need to create it
+    if (manual->curPage == NULL)// || ((page_t*)manual->curPage->val)->chapter != manual->chapter)
+    {
+        manual->curPage = paginateText(manual->markdown, &manual->pages);
+
+        if (reverse)
+        {
+            manual->curPage = manual->pages.last;
+        }
+    }
+}
+
+void manualEnterMode(display_t* disp)
+{
+    manual = calloc(1, sizeof(manual_t));
+
+    manual->disp = disp;
+
+    manual->fgColor = c555;
+    manual->bgColor = c000;
+
+    manual->navigate = RELOAD_PAGE;
+    manual->reload = true;
+
+    loadFont(MANUAL_FONT_UI, &manual->uiFont);
+    manual->bodyFont = manual->uiFont;
+    loadFont(MANUAL_FONT_HEADER, &manual->headerFont);
+
+    loadWsg("arrow12.wsg", &manual->arrowWsg);
+
+    markdownParams_t params =
+    {
+        .xMin = MANUAL_SIDE_MARGIN, .yMin = MANUAL_TOP_MARGIN,
+        .xMax = manual->disp->w - MANUAL_SIDE_MARGIN, .yMax = manual->disp->h - MANUAL_BOTTOM_MARGIN,
+        .bodyFont = &manual->bodyFont,
+        .headerFont = &manual->headerFont,
+        .color = manual->fgColor,
+    };
+
+    memcpy(&manual->mdParams, &params, sizeof(markdownParams_t));
+
+    manual->chapter = chapters;
+}
+
+void manualExitMode(void)
+{
+    freeFont(&manual->bodyFont);
+    freeFont(&manual->headerFont);
+
+    freeWsg(&manual->arrowWsg);
+
+    page_t* page;
+    while (NULL != (page = pop(&manual->pages)))
+    {
+        free(page->mdPos);
+        free(page);
+    }
+    // maybe redundant but oh well
+    clear(&manual->pages);
+
+    if (manual->markdown != NULL)
+    {
+        freeMarkdown(manual->markdown);
+    }
+
+    if (manual->text != NULL)
+    {
+        free(manual->text);
+    }
+    free(manual);
```

Comment:
You ran this with valgrind or libasan or something to check?

Reply:
Yeah, I'm generally always running with asan